### PR TITLE
convertToEs6Module: Avoid replacing entire function

### LINKED ISSUE #22492: Error replacing statement that shares line with another statement
**TypeScript Version:**  master

**Code**

See `refactorConvertToEs6Module_export_referenced.ts` (as of #22361)

**Expected behavior:**

```ts
_z; z;
```

**Actual behavior:**

```ts
_z  z;
```

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #22492
Will rebase once #22361 is in.

### CODE REVIEW

File: src/services/codefixes/convertToEs6Module.ts [hunk 1]
```diff
@@ -303,7 +300,27 @@ namespace ts.codefix {
         return makeExportDeclaration([createExportSpecifier(/*propertyName*/ undefined, "default")], moduleSpecifier);
     }
 
-    function convertExportsDotXEquals(name: string | undefined, exported: Expression): Statement {
+    function convertExportsDotXEquals({ left, right, parent }: BinaryExpression & { left: PropertyAccessExpression }, sourceFile: SourceFile, changes: textChanges.ChangeTracker): void {
```
Comment (amcasey) [inline src/services/codefixes/convertToEs6Module.ts]: Could "ExportsDotXEquals" be rephrased as "ExportsPropertyAssignment"?

File: src/services/codefixes/convertToEs6Module.ts [hunk 2]
```diff
@@ -303,7 +300,27 @@ namespace ts.codefix {
         return makeExportDeclaration([createExportSpecifier(/*propertyName*/ undefined, "default")], moduleSpecifier);
     }
 
-    function convertExportsDotXEquals(name: string | undefined, exported: Expression): Statement {
+    function convertExportsDotXEquals({ left, right, parent }: BinaryExpression & { left: PropertyAccessExpression }, sourceFile: SourceFile, changes: textChanges.ChangeTracker): void {
+        const name = left.name.text;
+        if ((isFunctionExpression(right) || isArrowFunction(right) || isClassExpression(right)) && (!right.name || right.name.text === name)) {
+            // `exports.f = function() {}` -> `export function f() {}` -- Replace `exports.f = ` with `export `, and insert the name after `function`.
+            changes.replaceRange(sourceFile, { pos: left.getStart(sourceFile), end: right.getStart(sourceFile) }, createToken(SyntaxKind.ExportKeyword), { suffix: " " });
```
Comment (amcasey) [inline src/services/codefixes/convertToEs6Module.ts]: Will this clobber trivia on the LHS of the assignment?

File: src/services/codefixes/convertToEs6Module.ts [hunk 3]
```diff
@@ -303,7 +300,27 @@ namespace ts.codefix {
...
+    function convertExportsDotXEquals({ left, right, parent }: BinaryExpression & { left: PropertyAccessExpression }, sourceFile: SourceFile, changes: textChanges.ChangeTracker): void {
+        const name = left.name.text;
+        if ((isFunctionExpression(right) || isArrowFunction(right) || isClassExpression(right)) && (!right.name || right.name.text === name)) {
+            // `exports.f = function() {}` -> `export function f() {}` -- Replace `exports.f = ` with `export `, and insert the name after `function`.
+            changes.replaceRange(sourceFile, { pos: left.getStart(sourceFile), end: right.getStart(sourceFile) }, createToken(SyntaxKind.ExportKeyword), { suffix: " " });
+
+            if (!right.name) changes.insertName(sourceFile, right, name);
+
+            const semi = findChildOfKind(parent, SyntaxKind.SemicolonToken, sourceFile);
+            if (semi) changes.deleteNode(sourceFile, semi, { useNonAdjustedEndPosition: true });
```
Comment (amcasey) [inline src/services/codefixes/convertToEs6Module.ts]: Will this clobber trailing trivia on the semicolon?

File: tests/cases/fourslash/refactorConvertToEs6Module_export_referenced.ts [hunk 4]
```diff
@@ -12,13 +12,14 @@
...
 ////exports.f = function(z) {
-////    z;
+////    exports.z; z;
 ////}
 
 // TODO: GH#22492 Should be a able access `exports.z` inside `exports.f`
 
 verify.codeFix({
     description: "Convert to ES6 module",
+    // TODO: GH#22492
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply tests/cases/fourslash/refactorConvertToEs6Module_export_referenced.ts]: :+1:

File: src/services/textChanges.ts [hunk 5]
```diff
@@ -477,6 +481,35 @@ namespace ts.textChanges {
...
 
+        public insertName(sourceFile: SourceFile, node: FunctionExpression | ClassExpression | ArrowFunction, name: string): void {
+            Debug.assert(!node.name);
+            if (node.kind === SyntaxKind.ArrowFunction) {
+                const arrow = findChildOfKind(node, SyntaxKind.EqualsGreaterThanToken, sourceFile)!;
+                const lparen = findChildOfKind(node, SyntaxKind.OpenParenToken, sourceFile);
+                if (lparen) {
+                    // `() => {}` --> `function f() {}`
+                    this.insertNodesAt(sourceFile, lparen.getStart(sourceFile), [createToken(SyntaxKind.FunctionKeyword), createIdentifier(name)], { joiner: " " });
+                    this.deleteNode(sourceFile, arrow);
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/textChanges.ts]: This is a bug already, see #23373

File: src/services/textChanges.ts [hunk 6]
```diff
@@ -477,6 +481,35 @@ namespace ts.textChanges {
             return Debug.failBadSyntaxKind(node); // We haven't handled this kind of node yet -- add it
         }
 
+        public insertName(sourceFile: SourceFile, node: FunctionExpression | ClassExpression | ArrowFunction, name: string): void {
+            Debug.assert(!node.name);
+            if (node.kind === SyntaxKind.ArrowFunction) {
+                const arrow = findChildOfKind(node, SyntaxKind.EqualsGreaterThanToken, sourceFile)!;
+                const lparen = findChildOfKind(node, SyntaxKind.OpenParenToken, sourceFile);
+                if (lparen) {
+                    // `() => {}` --> `function f() {}`
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/textChanges.ts]: See `if (node.body.kind !== SyntaxKind.Block)` below

File: src/services/textChanges.ts [hunk 7]
```diff
@@ -477,6 +481,35 @@ namespace ts.textChanges {
...
+                const arrow = findChildOfKind(node, SyntaxKind.EqualsGreaterThanToken, sourceFile)!;
+                const lparen = findChildOfKind(node, SyntaxKind.OpenParenToken, sourceFile);
+                if (lparen) {
+                    // `() => {}` --> `function f() {}`
+                    this.insertNodesAt(sourceFile, lparen.getStart(sourceFile), [createToken(SyntaxKind.FunctionKeyword), createIdentifier(name)], { joiner: " " });
+                    this.deleteNode(sourceFile, arrow);
+                }
+                else {
+                    // `x => {}` -> `function f(x) {}`
+                    this.insertNodesAt(sourceFile, first(node.parameters).getStart(sourceFile), [createToken(SyntaxKind.FunctionKeyword), createIdentifier(name + "(")], { joiner: " " });
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/textChanges.ts]: Good idea, when I first wrote this we didn't have `insertText`.

File: src/services/textChanges.ts [hunk 8]
```diff
@@ -477,6 +481,35 @@ namespace ts.textChanges {
...
+                    // Replaceing full range of arrow to get rid of the leading space -- replace ` =>` with `)`
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/textChanges.ts]: :+1:

File: src/services/textChanges.ts [hunk 9]
```diff
@@ -477,6 +481,35 @@ namespace ts.textChanges {
...
+                    this.replaceRange(sourceFile, arrow, createToken(SyntaxKind.CloseParenToken));
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/textChanges.ts]: #23373

File: src/services/codefixes/convertToEs6Module.ts (hunk 1 above)
Comment (andy-ms) [reply src/services/codefixes/convertToEs6Module.ts]: :+1:

File: src/services/codefixes/convertToEs6Module.ts (hunk 2 above)
Comment (andy-ms) [reply src/services/codefixes/convertToEs6Module.ts]: #23373

File: src/services/codefixes/convertToEs6Module.ts (hunk 3 above)
Comment (andy-ms) [reply src/services/codefixes/convertToEs6Module.ts]: #23373

**Pull request merged**